### ISSUE TITLE
Allow path-like objects in `fnmatch.filter` on POSIX platforms

### ISSUE THREAD

**Issue by picnixz** [python/cpython#123135]
In `fnmatch.filter`, it says that

```py
# normcase on posix is NOP. Optimize it away from the loop.
```

This is not the case. `os.path.normcase` actually calls `os.fspath` on POSIX, meaning that path-like objects are allowed. Thus, we cannot optimize it away from the loop. I'll update #123122 accordingly.

Related: https://github.com/python/cpython/pull/123122#issuecomment-2295320508

### CPython versions tested on:

CPython main branch

### Linked PRs
* gh-123122

**Comment by picnixz** [comment]
Marking this one as a feature instead of a bugfix (hence it will only be included in 3.14 and not backported). See https://github.com/python/cpython/pull/123122#discussion_r1724071890.

**Comment by picnixz** [comment]
See https://github.com/python/cpython/issues/123215#issuecomment-2308763569 for the rationale of closing this issue as `not planned` for now.
